File: 3348-smallest-divisible-digit-product-ii/3348-smallest-divisible-digit-product-ii.py
class Solution:
    def smallestNumber(self, num: str, t: int) -> str:
        # Step 1: Prime factorize t into 2, 3, 5, 7
        temp_t = t
        counts = {2: 0, 3: 0, 5: 0, 7: 0}
        for p in [2, 3, 5, 7]:
            while temp_t % p == 0:
                counts[p] += 1
                temp_t //= p
        
        # If t has prime factors > 7, it's impossible
        if temp_t > 1:
            return "-1"

        def get_min_digits(c2: int, c3: int, c5: int, c7: int) -> int:
            """Calculates minimum count of non-zero digits needed to satisfy factor counts."""
            c2, c3, c5, c7 = max(0, c2), max(0, c3), max(0, c5), max(0, c7)
            
            d9, r3 = divmod(c3, 2)
            d8, r2 = divmod(c2, 3)
            
            d6 = 0
            if r3 == 1 and r2 == 1:
                d6 = 1
                r3, r2 = 0, 0
            elif r3 == 1 and r2 == 2:
                d6 = 1
                r3, r2 = 0, 1
            
            d4, r2 = divmod(r2, 2)
            return d9 + d8 + d6 + d4 + r3 + r2 + c5 + c7

        def get_suffix(c2: int, c3: int, c5: int, c7: int, target_len: int) -> str:
            """Builds lexicographically smallest valid suffix of target_len."""
            c2, c3, c5, c7 = max(0, c2), max(0, c3), max(0, c5), max(0, c7)
            
            min_len = get_min_digits(c2, c3, c5, c7)
            ones = target_len - min_len
            
            d9, r3 = divmod(c3, 2)
            d8, r2 = divmod(c2, 3)
            
            d6 = 0
            if r3 == 1 and r2 == 1:
                d6 = 1
                r3, r2 = 0, 0
            elif r3 == 1 and r2 == 2:
                d6 = 1
                r3, r2 = 0, 1
                
            d4, r2 = divmod(r2, 2)
            
            digits = (
                ['2'] * r2 +
                ['3'] * r3 +
                ['4'] * d4 +
                ['5'] * c5 +
                ['6'] * d6 +
                ['7'] * c7 +
                ['8'] * d8 +
                ['9'] * d9
            )
            return "1" * ones + "".join(digits)

        # Factor contributions for digits '1' through '9'
        digit_factors = {
            '1': (0, 0, 0, 0), '2': (1, 0, 0, 0), '3': (0, 1, 0, 0),
            '4': (2, 0, 0, 0), '5': (0, 0, 1, 0), '6': (1, 1, 0, 0),
            '7': (0, 0, 0, 1), '8': (3, 0, 0, 0), '9': (0, 2, 0, 0)
        }

        n = len(num)
        first_zero = num.find('0')
        limit = first_zero if first_zero != -1 else n
        
        # Track required factors after prefix matching
        pref_c2, pref_c3, pref_c5, pref_c7 = counts[2], counts[3], counts[5], counts[7]
        prefix_reqs = [(pref_c2, pref_c3, pref_c5, pref_c7)]
        
        for i in range(limit):
            f2, f3, f5, f7 = digit_factors[num[i]]
            pref_c2 -= f2
            pref_c3 -= f3
            pref_c5 -= f5
            pref_c7 -= f7
            prefix_reqs.append((pref_c2, pref_c3, pref_c5, pref_c7))

        # Check if num itself is valid
        if first_zero == -1 and get_min_digits(*prefix_reqs[n]) == 0:
            return num

        # Step 2: Try changing digit at index i to a larger digit
        for i in range(limit, -1, -1):
            if i == n:
                continue
            
            c2, c3, c5, c7 = prefix_reqs[i]
            start_digit = int(num[i]) + 1 if i < limit else 1
            
            for d in range(start_digit, 10):
                d_str = str(d)
                f2, f3, f5, f7 = digit_factors[d_str]
                rem_c2, rem_c3 = c2 - f2, c3 - f3
                rem_c5, rem_c7 = c5 - f5, c7 - f7
                
                rem_len = n - 1 - i
                if get_min_digits(rem_c2, rem_c3, rem_c5, rem_c7) <= rem_len:
                    prefix = num[:i] + d_str
                    suffix = get_suffix(rem_c2, rem_c3, rem_c5, rem_c7, rem_len)
                    return prefix + suffix

        # Step 3: Expand to length n + 1 if same length is impossible
        target_len = max(n + 1, get_min_digits(counts[2], counts[3], counts[5], counts[7]))
        return get_suffix(counts[2], counts[3], counts[5], counts[7], target_len)

File: 3348-smallest-divisible-digit-product-ii/3348-smallest-divisible-digit-product-ii.java
class Solution {
    public String smallestNumber(String num, long t) {
        // Step 1: Factorize t into prime factors 2, 3, 5, 7
        long tempT = t;
        int[] counts = new int[10];
        int[] primes = {2, 3, 5, 7};
        for (int p : primes) {
            while (tempT % p == 0) {
                counts[p]++;
                tempT /= p;
            }
        }

        // If t has prime factors > 7, it's impossible to form the product
        if (tempT > 1) {
            return "-1";
        }

        int n = num.length();
        int firstZero = num.indexOf('0');
        int limit = (firstZero != -1) ? firstZero : n;

        // Track required factors after prefix matching
        int[][] prefixReqs = new int[limit + 1][10];
        prefixReqs[0][2] = counts[2];
        prefixReqs[0][3] = counts[3];
        prefixReqs[0][5] = counts[5];
        prefixReqs[0][7] = counts[7];

        int[][] digitFactors = new int[10][10];
        digitFactors[2][2] = 1;
        digitFactors[3][3] = 1;
        digitFactors[4][2] = 2;
        digitFactors[5][5] = 1;
        digitFactors[6][2] = 1; digitFactors[6][3] = 1;
        digitFactors[7][7] = 1;
        digitFactors[8][2] = 3;
        digitFactors[9][3] = 2;

        for (int i = 0; i < limit; i++) {
            int d = num.charAt(i) - '0';
            for (int p : primes) {
                prefixReqs[i + 1][p] = prefixReqs[i][p] - digitFactors[d][p];
            }
        }

        // Check if num itself is valid
        if (firstZero == -1 && getMinDigits(prefixReqs[n][2], prefixReqs[n][3], prefixReqs[n][5], prefixReqs[n][7]) == 0) {
            return num;
        }

        // Step 2: Try changing digit at index i to a larger digit
        for (int i = limit; i >= 0; i--) {
            if (i == n) continue;

            int c2 = prefixReqs[i][2];
            int c3 = prefixReqs[i][3];
            int c5 = prefixReqs[i][5];
            int c7 = prefixReqs[i][7];

            int startDigit = (i < limit) ? (num.charAt(i) - '0' + 1) : 1;

            for (int d = startDigit; d <= 9; d++) {
                int remC2 = c2 - digitFactors[d][2];
                int remC3 = c3 - digitFactors[d][3];
                int remC5 = c5 - digitFactors[d][5];
                int remC7 = c7 - digitFactors[d][7];

                int remLen = n - 1 - i;
                if (getMinDigits(remC2, remC3, remC5, remC7) <= remLen) {
                    StringBuilder sb = new StringBuilder();
                    sb.append(num, 0, i).append(d);
                    sb.append(getSuffix(remC2, remC3, remC5, remC7, remLen));
                    return sb.toString();
                }
            }
        }

        // Step 3: Expand length if same length is impossible
        int minNeeded = getMinDigits(counts[2], counts[3], counts[5], counts[7]);
        int targetLen = Math.max(n + 1, minNeeded);
        return getSuffix(counts[2], counts[3], counts[5], counts[7], targetLen);
    }

    private int getMinDigits(int c2, int c3, int c5, int c7) {
        c2 = Math.max(0, c2);
        c3 = Math.max(0, c3);
        c5 = Math.max(0, c5);
        c7 = Math.max(0, c7);

        int d9 = c3 / 2, r3 = c3 % 2;
        int d8 = c2 / 3, r2 = c2 % 3;

        int d6 = 0;
        if (r3 == 1 && r2 == 1) {
            d6 = 1; r3 = 0; r2 = 0;
        } else if (r3 == 1 && r2 == 2) {
            d6 = 1; r3 = 0; r2 = 1;
        }

        int d4 = r2 / 2;
        r2 %= 2;

        return d9 + d8 + d6 + d4 + r3 + r2 + c5 + c7;
    }

    private String getSuffix(int c2, int c3, int c5, int c7, int targetLen) {
        c2 = Math.max(0, c2);
        c3 = Math.max(0, c3);
        c5 = Math.max(0, c5);
        c7 = Math.max(0, c7);

        int minLen = getMinDigits(c2, c3, c5, c7);
        int ones = targetLen - minLen;

        int d9 = c3 / 2, r3 = c3 % 2;
        int d8 = c2 / 3, r2 = c2 % 3;

        int d6 = 0;
        if (r3 == 1 && r2 == 1) {
            d6 = 1; r3 = 0; r2 = 0;
        } else if (r3 == 1 && r2 == 2) {
            d6 = 1; r3 = 0; r2 = 1;
        }

        int d4 = r2 / 2;
        r2 %= 2;

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < ones; i++) sb.append('1');
        for (int i = 0; i < r2; i++) sb.append('2');
        for (int i = 0; i < r3; i++) sb.append('3');
        for (int i = 0; i < d4; i++) sb.append('4');
        for (int i = 0; i < c5; i++) sb.append('5');
        for (int i = 0; i < d6; i++) sb.append('6');
        for (int i = 0; i < c7; i++) sb.append('7');
        for (int i = 0; i < d8; i++) sb.append('8');
        for (int i = 0; i < d9; i++) sb.append('9');

        return sb.toString();
    }
}